design/design-doc-smaller-providers.md

```diff
@@ -0,0 +1,704 @@
+theme was that community members don’t want to manage more ProviderConfigs than
+they do today. A community member running 10 smaller AWS providers, each with
+MRs spanning 10 AWS accounts would go from 10 ProviderConfigs to 10 x 10 = 100
+ProviderConfigs.
+
+ProviderConfigs are typically created by humans, or as part of a Composition. MR
+controllers create a ProviderConfigUsage for each MR. A ProviderConfig
+controller places a finalizer on each ProviderConfig to prevent it from being
+deleted while ProviderConfigUsages reference it.
+
+I propose that as part of breaking up the biggest providers, each provider
+"family" include a special “config” provider - e.g. `provider-aws-config`, or
+`provider-azure-config`. This config provider would:
+
+* Install the ProviderConfig, ProviderConfigUsage, and StoreConfig CRDs.
+* Start a controller to prevent the ProviderConfig being deleted while
+  ProviderConfigUsages referenced it.
+
+All other providers in the provider family would take a package manager
+dependency on the relevant config provider. This:
+
+* Ensures the config provider is installed automatically.
+* Reduces coupling - dependent providers can depend on a semantic version range.
+
+Consider the following scenarios:
+
+* A ProviderConfig is updated with a new field.
+* A ProviderConfig enum field is updated with a new value
+
+The first scenario is well suited to semantic version dependencies. The new
+field would need to be optional to avoid a breaking schema change to the
+ProviderConfig type (this is true today). Providers depending on “at least
+version X” of the config provider would be unaffected by updates; they would
+simply ignore the new optional field. Providers that wished to leverage the new
+field would update their dependency to require at least the version in which the
+field was introduced.
+
+The second scenario is a little trickier. If the config provider was updated and
+existing ProviderConfigs were updated in place to use the new enum value (e.g.
+`spec.credentials.source: Vault`) all dependent providers would need to be
+updated. Any that weren’t would treat the new enum value as unknown. In this
+scenario the best practice would be to either update all providers with support
+for the new value, or update only some providers and have them use a new
+ProviderConfig using the new value, leaving the existing ProviderConfigs
+untouched.
+
+### Inter-Provider Dependencies
+
+As established in the [Provider Configs](#provider-configs) section, all
+providers within a provider family will depend on a single config provider. For
+example all `provider-aws-*` providers will have a package dependency on
+`provider-aws-config`.
+
+A common misconception is that providers will need to depend on the providers
+they may (cross-resource) reference. This is not the case. I expect that the
+_only_ dependency providers within a family will have is on their config
```

Comment:
Another exception seems to be recently added advanced to  `core`, e.g. `provider-aws-ec2` depends on `provider-aws-ec2-core`

Reply:
Yes good catch. I'll update after I chat with @bassam to get his perspective on whether we should break up larger providers later today.

Reply:
this is another reason to hold of on Large Services for now. There is a natural isolation between two independent services, but if we break the large services further we end up with arbitrary packaging of core vs. non-core and the dependencies might not line up as cleanly.